src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs

```diff
@@ -50,4 +53,29 @@ public override bool Equals(BoundDagEvaluation obj)
                 this.Index == ((BoundDagIndexEvaluation)obj).Index;
         }
     }
+
+    partial class BoundDagIndexerEvaluation
+    {
+        public override int GetHashCode() => base.GetHashCode() ^ this.Index;
+        public override bool Equals(BoundDagEvaluation obj)
+        {
+            return this == obj ||
+                base.Equals(obj) &&
+                // base.Equals checks the kind field, so the following cast is safe
+                this.Index == ((BoundDagIndexerEvaluation)obj).Index;
+        }
+    }
+
+    partial class BoundDagSliceEvaluation
+    {
+        public override int GetHashCode() => base.GetHashCode() ^ this.StartIndex ^ this.EndIndex;
+        public override bool Equals(BoundDagEvaluation obj)
+        {
+            return this == obj ||
+                base.Equals(obj) &&
+                // base.Equals checks the kind field, so the following cast is safe
+                (BoundDagSliceEvaluation)obj is var e &&
```

Comment:
What if `obj` is not a `BoundDagSliceEvaluation`?
This should probably be `obj is BoundDagSliceEvaluation e &&` instead.

Similar comment applies to `BoundDagIndexerEvaluation` above (line 65).

Reply:
>`// base.Equals checks the kind field, so the following cast is safe` ;)